### PR TITLE
Migrating to typescript

### LINKED ISSUE #43: Migrate to Typescript
Migrate project to Typescript. Project configuration, style, etc, should be consistent with [sqs-consumer](https://github.com/bbc/sqs-consumer)

### PR DESCRIPTION
**Pull request by jeanrauwers**
fixes #43

### CODE REVIEW

File: src/producer.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import { SQS } from 'aws-sdk';
+import { entryFromMessage } from './types';
+const requiredOptions = [
+  'queueUrl'
+];
+
+interface Options {
```
Comment (nspragg) [inline src/producer.ts:7]: Consider renaming: `ProducerOptions` is more descriptive

File: src/producer.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  queueUrl?: string;
+  batchSize?: number;
+  sqs?: SQS;
+  region?: string;
+}
+
+interface ProducerInterface {
```
Comment (nspragg) [inline src/producer.ts:14]: We probably don't need this interface, as there is only one class. It's unlikely that we'll need multiple implementations of the Producer at a given time.

File: src/producer.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+      if (!options[option]) {
+        throw new Error(`Missing SQS producer option [${option}].`);
+      }
+    }
+    if (options.batchSize > 10 || options.batchSize < 1) {
+      throw new Error('SQS batchSize option must be between 1 and 10.');
+    }
+  }
+
+  async _sendBatch(failedMessages?: string[], messages?: string[], startIndex?: number): Promise<string[]> {
```
Comment (nspragg) [inline src/producer.ts:49]: Consider using the private modifier and renaming the method to remove the '_'

File: src/producer.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    this.validate(options);
+    this.queueUrl = options.queueUrl;
+    this.batchSize = options.batchSize || 10;
+    this.sqs = options.sqs || new SQS({
+      ...options,
+      region: options.region || 'eu-west-1'
+    });
+  }
+
+  validate(options: Options): void {
```
Comment (nspragg) [inline src/producer.ts:38]: Consider using the private modifier

File: src/types.ts [hunk 5]
```diff
@@ -1,19 +1,28 @@
...
+        throw new Error(`Object messages must have 'id' prop`);
     }
 
     if (message.deduplicationId && !message.groupId) {
-        throw new Error("FIFO Queue messages must have 'groupId' prop");
+        throw new Error(`FIFO Queue messages must have 'groupId' prop`);
     }
 
-    const entry = {
+    const entry: any = {
```
Comment (nspragg) [inline src/types.ts:25]: This looks like an AWS defined type, e.g `SendMessageBatchRequestEntry` (this needs confirming). Consider using predefined types where possible.

File: src/types.ts [hunk 6]
```diff
@@ -64,14 +71,14 @@ function entryFromObject(message) {
     return entry;
 }
 
-function entryFromString(message) {
+function entryFromString(message: Message): any {
```
Comment (nspragg) [inline src/types.ts:74]: The appropriate aws entry type can be returned from this function.

File: src/types.ts [hunk 7]
```diff
@@ -64,14 +71,14 @@ function entryFromObject(message) {
...
     return {
         Id: message,
         MessageBody: message
     };
 }
 
-function entryFromMessage(message) {
+export function entryFromMessage(message: any): string[] {
```
Comment (nspragg) [inline src/types.ts:81]: The appropriate aws entry type can be returned from this function; should return the same type as `entryFromString` and `entryFromObject`, refactored.

File: src/validation.ts [hunk 8]
```diff
@@ -1,13 +1,13 @@
...
     return typeof value === 'string' || value instanceof String;
 }
 
-function isObject(value) {
+function isObject(value: any): boolean {
     return value && typeof value === 'object' && value instanceof Object;
 }
 
-function isMessageAttributeValid(messageAttribute) {
+function isMessageAttributeValid(messageAttribute: any): any {
```
Comment (nspragg) [inline src/validation.ts:10]: Technically, this function returns boolean. Although in practice reads "always true or throw Error". This needs refactoring but can be done when the basic TS conversion is complete.

File: src/producer.ts (hunk 2 above)
Comment (jeanrauwers) [reply src/producer.ts:14]: agreed

File: src/validation.ts (hunk 8 above)
Comment (jeanrauwers) [reply src/validation.ts:10]: agreed!

File: test/producer.ts [hunk 9]
```diff
@@ -22,19 +22,19 @@ describe('Producer', function () {
     });
   });
 
-  afterEach(function () {
+  afterEach(() => {
     sqs.sendMessageBatch.restore();
   });
 
-  async function rejects(promise, errMessage) {
+  async function rejects(promise: any, errMessage: any): Promise<any> {
```
Comment (nspragg) [inline test/producer.ts:29]: These `any` types need to be refactored. For example, `rejects` could be:
```ts
  async function rejects(producerResponse: Promise<string[]>, errMessage: string): Promise<void> {
    let thrown = false;
    try {
      await producerResponse;
    } catch (err) {
      thrown = true;
      assert.equal(err.message, errMessage);
    }
    if (!thrown) { assert.fail(`Should have thrown: ${errMessage}`); }
  }
```

**Pull request merged**